features/filesystem/storagelite/source/StorageLite.cpp

```diff
@@ -0,0 +1,1846 @@
+            return ret;
+        }
+        if (!(flags & rollback_protect_flag)) {
+            continue;
+        }
+        os_ret = nvstore.remove(nvstore_key);
+        if ((os_ret != NVSTORE_SUCCESS) && (os_ret != NVSTORE_NOT_FOUND)) {
+            return STORAGELITE_OS_ERROR;
+        }
+    }
+    return ret;
+}
+
+
+int StorageLite::init(BlockDevice *bd, size_t max_files)
+{
+    ram_table_entry_t *ram_table;
+    area_state_e area_state[STORAGELITE_NUM_AREAS];
+    uint32_t init_attempts_val;
+    uint32_t next_offset;
+    uint32_t flags, hash;
+    uint32_t actual_data_size;
+    int os_ret;
+    int ret = STORAGELITE_SUCCESS;
+    uint16_t versions[STORAGELITE_NUM_AREAS];
+    uint16_t actual_name_size;
+    uint16_t nvstore_key;
+
+    if (_init_done) {
+        // Null BD signals the use of current one
+        if (!bd || (bd == _user_bd)) {
+            return STORAGELITE_SUCCESS;
+        }
+        // Reinitializing with a different BD - need to deinit first
+        deinit();
+    }
+    MBED_ASSERT(bd);
+
+    // This handles the case that init function is called by more than one thread concurrently.
+    // Only the one who gets the value of 1 in _init_attempts_val will proceed, while others will
+    // wait until init is finished.
+    init_attempts_val = core_util_atomic_incr_u32(&_init_attempts, 1);
+    if (init_attempts_val != 1) {
+        while (!_init_done) {
+            wait_ms(1);
+        }
+        return STORAGELITE_SUCCESS;
+    }
+
+    _max_files = max_files;
+    ram_table = new ram_table_entry_t[_max_files];
+    _ram_table = ram_table;
+    _num_ram_table_entries = 0;
+
+    NVStore& nvstore = NVStore::get_instance();
+    os_ret = nvstore.init();
+    MBED_ASSERT(!os_ret);
+
+    uint16_t curr_max_keys = nvstore.get_max_keys();
+
+    if ((size_t)(curr_max_keys - NVSTORE_NUM_PREDEFINED_KEYS) < max_files) {
+        // Need to have sufficient number of NVStore keys for rollback protection
+        // FIXME: Is this the correct way to do that?
+        nvstore.set_max_keys(max_files + NVSTORE_NUM_PREDEFINED_KEYS);
+        os_ret = nvstore.init();
+        MBED_ASSERT(!os_ret);
+    }
+
+    if (!_mutex) {
+        _mutex = new PlatformMutex;
+    }
+
+    bd->init();
+
+    // Underlying BD size must fit into 32 bits
+    MBED_ASSERT((uint32_t)bd->size() == bd->size());
+
+    // Underlying BD must have flash attributes, i.e. have an erase value
+    MBED_ASSERT(bd->get_erase_value() != -1);
+
+    _size = (size_t) -1;
+    _user_bd = bd;
+
+    // If the BD's program (or read) size is larger than header size, then this would make our
+    // read_record and write_record code much more complex (as we'll have to read/write chunks
+    // of the data along with the header). If this is the case, use the buffered block device
+    // to reduce the underlying BD read and program sizes
+    if (bd->get_program_size() <= sizeof(record_header_t)) {
```

Comment:
Agree about checking the header size as a multiple of the program size. Off topic, I think block devices should make sure read/program/erase sizes are powers of 2. This would make that check redundant.

Reply:
I'm not sure I understand, that only helps here if `sizeof(record_header_t)` is also a power of 2 doesn't it?

Reply:
I should also point that out in the comments - header is 32 bytes, and any change should keep it a power of 2.

Reply:
Anyway fixed to check whether it's a multiple.